src/Form/FormMapper.php

```diff
@@ -130,7 +130,10 @@ public function add($name, $type = null, array $options = [], array $fieldDescri
             $name = $fieldDescription->getName();
 
             // Note that the builder var is actually the formContractor:
-            $options = array_replace_recursive($this->builder->getDefaultOptions($type, $fieldDescription) ?? [], $options);
+            $options = array_replace_recursive(
+                $this->builder->getDefaultOptions($type, $fieldDescription, $options),
```

Comment:
Probably I'm missing something, but why do we need to pass the form options to the builder (persistence bundles) if we already have them already here?

Reply:
Because they are transforme.

From
```
$options = ['by_reference' => true];
```
To
``` 
$options = [`type_options` => ['by_reference' => true]]
```
(And multiple others modifications)

And the options passed are not always the same ; the depends on the `type`, etc

Reply:
I remember some unintended behavior regarding these options (#6078).

Reply:
Not sure to understand. ^^'
Is it an answer to franmomu ? A request change to me ?

Reply:
Sorry. The comment is regarding the modifications you've mentioned about the options.

Reply:
I still don't understand why it's an issue.

Previously it was
```
$options = array_replace_recursive(
                $this->builder->getDefaultOptions($type, $fieldDescription),
                $options
)
```

So $options was updated from
```
$options = ['foo' => 'bar' , 'by_reference' => false, ... ]
```
To
```
$options = ['foo' => 'bar' , 'by_reference' => false, ..., 'type' => AdminType::class, 'type_options' => [ ... ]. ]
```

Now, it's 
```
$options = array_replace_recursive(
                $this->builder->getDefaultOptions($type, $fieldDescription, $options),
                $options
)
```

So $options is updated from
```
$options = ['foo' => 'bar' , 'by_reference' => false, ... ]
```
To
```
$options = ['foo' => 'bar' , 'by_reference' => false, ..., 'type' => AdminType::class, 'type_options' => [ ..., 'by_reference' => false ]. ]
```

Franmomu was asking why the options is passed to DoctrineORM when it could be done in SonataAdmin.
But seems sometime the `getDefaultOptions` returns some `type_options`, sometimes there is not, I think it's better to do this in the persistence bundle (and it's already done).

Reply:
I'm not saying there is an issue with your changes. I'm pointing out that there is already some unexpected modifications in the options, related to the usage of `array_merge_recursive()` as mentioned in the linked PR https://github.com/sonata-project/SonataAdminBundle/pull/6078#issue-412703816.

Reply:
What I meant is... is there a difference between this code and the modified one in persistence bundles and:

```php
$formOptions = $options;
$options = array_replace_recursive($this->builder->getDefaultOptions($type, $fieldDescription) ?? [], $formOptions);

if (isset($options['type_options') && isset($formOptions['by_reference'])) {
    $options['type_options']['by_reference'] = $formOptions['by_reference'];
}     
```

which means to leave persistence bundles as they were (less duplicated code and less to maintain) and not adding a conflict with them.

In fact looking at [`FormContractor:: getDefaultOptions`](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/711eba35e0bef87ab57fa5d4a8ee8b5f6457a62f/src/Builder/FormContractor.php#L99), looks that could be done in the admin bundle since the only persistence-related reference I see is to `ClassMetadata` and that could be abstracted in the `FieldDescriptionInterface`.

Reply:
> What I meant is... is there a difference between this code and the modified one in persistence bundles and:
> 
> ```
> $formOptions = $options;
> $options = array_replace_recursive($this->builder->getDefaultOptions($type, $fieldDescription) ?? [], $formOptions);
> 
> if (isset($options['type_options') && isset($formOptions['by_reference'])) {
>     $options['type_options']['by_reference'] = $formOptions['by_reference'];
> }     
> ```

IMHO, since the persistence bundle are setting some options in `type_options`, everything should be done in it.
Maybe some will need to set the `by_reference` and some won't. We can't say ; only the persistence bundle will decide and we shouldn't guess in the SonataAdminBundle.

> In fact looking at FormContractor:: getDefaultOptions, looks that could be done in the admin bundle since the only persistence-related reference I see is to ClassMetadata and that could be abstracted in the FieldDescriptionInterface.

This PR is a patch for a bug existing for 1 month. Can we merge it ? 
A refacto to move all the logic from persistence bundle to admin bundle is certainly a good idea but it will take times.

Reply:
If the refactoring is the better choice we have, I think we should create an issue to track the required changes and update the `NEXT_MAJOR` comment in order to avoid pointing the goal to an approach that we consider slack.

Reply:
> If the refactoring is the better choice we have, I think we should create an issue to track the required changes and update the `NEXT_MAJOR` comment in order to avoid pointing the goal to an approach that we consider slack.

If I understand correctly, @franmomu would like to deprecate `getDefaultOptions` since
https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/src/Builder/FormContractor.php#L99
https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/blob/3.x/src/Builder/FormContractor.php#L97
are similar.

But if you look at other bundle, like 
https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/blob/2.x/src/Builder/FormContractor.php#L118
That's less similar. 

So I'm not sure the issue should be resolved.
We will have to live with similar code between ORM and MongoDB, but if we want to allow other persistence bundle, that's the right way to do it.

If @franmomu see things differently, I invite him to create the issue to explain the strategy.

But this debate should not block this PR.

Reply:
The reasoning seems fine to me.
I just have a concern. Since we are adding a 3rd argument in order to fix this issue properly in the next major, shouldn't be better to provide the fix by using this new argument through `func_get_args()` and deprecate the case where it's considered required and not used?
I didn't see how this new argument will be used in the method, so in case my previous concern can not be done, I think at least we should leave the comment with the code fragment that must be uncommented in the next major.

Reply:
I don't understand @phansys.

To me, the only NEXT_MAJOR comment needed is https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/src/Builder/FormContractor.php#L101

The 3rd argument will always be used from now, so I don't understand why you want to deprecate it ?
This is a long term fix.

Reply:
>The 3rd argument will always be used from now, so I don't understand why you want to deprecate it ?

I'm not saying we should deprecate the argument, please re-read my comment:

>deprecate the case where it's considered required and not used

Reply:
>To me, the only NEXT_MAJOR comment needed is https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/src/Builder/FormContractor.php#L101

Sorry, didn't see that the call to `func_get_args()` and the `NEXT_MAJOR` comment were already there before this PR.

I think then we should just trigger a deprecation if the `FormContractor::getDefaultOptions()` method is called without the 3rd argument. Do you think it could be done in this PR?

Reply:
The getDefaultOptions is implemented in persistence bundle, so I can't do it in this SonataAdmin PR.

But I can create PR on persistence bundle then.